File: WOT-iOS/WOTPivot/WOTPivot/core/adapters/webRequest/WOTRequestArguments.m
//
//  WOTRequestArguments.m
//  WOTPivot
//
//  Created by Pavel Yeshchyk on 3/2/20.
//  Copyright © 2020 Pavel Yeshchyk. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "WOTRequestArguments.h"
#import <WOTPivot/WOTPivot-Swift.h>

@interface WOTRequestArguments ()
@property (nonatomic, strong) NSMutableDictionary *dict;
@end

@implementation WOTRequestArguments

- (id)init:(NSDictionary *)dictionary {
    
    self = [super init];
    if (self) {
        [dictionary.allKeys enumerateObjectsUsingBlock:^(id  _Nonnull key, NSUInteger idx, BOOL * _Nonnull stop) {
            NSString *plainarray = dictionary[key];
            NSArray *arr = [plainarray componentsSeparatedByString:@","];
            [self setValues:arr forKey:key];
        }];
    }
    return self;
}

- (void)setValues:(NSArray *)values forKey:(NSString *)key {
    if (_dict == nil) {
        _dict = [[NSMutableDictionary alloc] init];
    }
    _dict[key] = values;
}

- (NSDictionary *)asDictionary {
    return [_dict copy];
}

- (NSString *)escapedValueForKey:(NSString *)key {
    NSMutableArray *result = [[NSMutableArray alloc] init];
    [_dict[key] enumerateObjectsUsingBlock:^(NSString *  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        NSString *objString = (NSString *)obj;
        NSString *string = [objString urlEncoded];
        [result addObject:string];
    }];
    return [result componentsJoinedByString:@","];
}

- (NSString *)composeQuery {
    NSMutableArray *result = [[NSMutableArray alloc] init];
    [_dict.allKeys enumerateObjectsUsingBlock:^(id  _Nonnull key, NSUInteger idx, BOOL * _Nonnull stop) {
        NSString *escapedValue = [self escapedValueForKey:key];
        [result addObject: [NSString stringWithFormat:@"%@=%@",key, escapedValue] ];
    }];
    return [result componentsJoinedByString:@"&"];
}

@end

File: WOT-iOS/WOTPivot/WOTPivot/core/adapters/webRequest/WOTRequestArguments.swift
//
//  WOTRequestArguments.swift
//  WOTPivot
//
//  Created by Pavel Yeshchyk on 3/4/20.
//  Copyright © 2020 Pavel Yeshchyk. All rights reserved.
//

import Foundation

@objc
public protocol WOTRequestArgumentsProtocol {
    @objc
    init(_ dictionary: JSON)

    @objc
    func setValues(_ values: Any, forKey: AnyHashable)

    @objc
    func buildQuery(_ custom: JSON) -> String
}

@objc
open class WOTRequestArguments: NSObject, WOTRequestArgumentsProtocol {
    @objc
    public private(set) var dictionary = JSON()

    @objc
    public override var description: String {
        return "\(dictionary)"
    }

    @objc
    required convenience public init(_ dictionary: JSON) {
        self.init()

        dictionary.keys.forEach {
            if let plainArray = dictionary[$0] as? String {
                let joinedString = plainArray.components(separatedBy: ",")
                self.dictionary[$0] = joinedString
            }
        }
    }

    @objc
    public func setValues(_ values: Any, forKey: AnyHashable) {
        dictionary[forKey] = values
    }

    @objc
    public func buildQuery(_ custom: JSON = [:]) -> String {
        var mixture = custom
        mixture.append(with: dictionary)
        return mixture.asURLQueryString()
    }

    @objc
    open override var hash: Int {
        do {
            let data = try JSONSerialization.data(withJSONObject: dictionary, options: JSONSerialization.WritingOptions.prettyPrinted)

            var hasher = Hasher()
            hasher.combine(data)
            return hasher.finalize()

        } catch {
            return 0
        }
    }
}
